betty/job/queue.py

```diff
@@ -0,0 +1,234 @@
+"""
+The job queue API.
+"""
+
+from __future__ import annotations
+
+from collections import defaultdict
+from enum import IntEnum
+from graphlib import TopologicalSorter
+from typing import (
+    TYPE_CHECKING,
+    Generic,
+    Self,
+    TypeAlias,
+    TypeVar,
+    final,
+)
+
+from betty.concurrent import AsynchronizedLock
+from betty.job import Context
+from betty.typing import internal
+
+if TYPE_CHECKING:
+    from collections.abc import (
+        AsyncIterator,
+        Awaitable,
+        Callable,
+        MutableMapping,
+        MutableSequence,
+    )
+    from types import TracebackType
+
+_ContextT = TypeVar("_ContextT", bound=Context)
+Job: TypeAlias = "Callable[[Queue[_ContextT]], Awaitable[None]]"
+
+
+@internal
+class StopWorking(Exception):
+    """
+    Raised when a queue has stopped working.
+    """
+
+
+@internal
+@final
+class Cancelled(StopWorking):
+    """
+    Raised when a queue has cancelled.
+    """
+
+
+@internal
+@final
+class Completed(StopWorking):
+    """
+    Raised when a queue has completed.
+    """
+
+
+class Duration(IntEnum):
+    """
+    A duration indication.
+    """
+
+    NEGLIGIBLE = 0
+    #: Less than a second.
+    SECONDS = 1
+    #: More than a second and less than a minute.
+    MINUTES = 2
+    #: More than a minute.
+
+
+@internal
+@final
```

Comment:
When returning batches, we don't want them all to be of the highest duration. Do it tiered, e.g. return at most one job that takes minutes. Also what if a worker gets one long job and many short ones. It will then spend most of its time just working on the long job. Not a problem I'm with async, but more so in the future with threading and multiprocessing.